riscv/debug_module.cc

```diff
@@ -715,120 +730,120 @@ bool debug_module_t::perform_abstract_command()
 
-      } else if (regno >= 0x1020 && regno < 0x1040 && config.support_abstract_fpr_access) {
-        unsigned fprnum = regno - 0x1020;
+      switch (size) {
+        case 2:
+          if (write)
+            write32(debug_abstract, i++, lw(regnum, ZERO, debug_data_start));
+          else
+            write32(debug_abstract, i++, sw(regnum, ZERO, debug_data_start));
+          break;
+        case 3:
+          if (write)
+            write32(debug_abstract, i++, ld(regnum, ZERO, debug_data_start));
+          else
+            write32(debug_abstract, i++, sd(regnum, ZERO, debug_data_start));
+          break;
+        default:
+          abstractcs.cmderr = CMDERR_NOTSUP;
+          return true;
+      }
 
-        if (write) {
-          switch (size) {
-            case 2:
-              write32(debug_abstract, i++, flw(fprnum, ZERO, debug_data_start));
-              break;
-            case 3:
-              write32(debug_abstract, i++, fld(fprnum, ZERO, debug_data_start));
-              break;
-            default:
-              abstractcs.cmderr = CMDERR_NOTSUP;
-              return true;
-          }
+      if (regno == 0x1000 + S0 && write) {
+        /*
+         * The exception handler starts out be restoring dscratch to s0,
+         * which was saved before executing the abstract memory region. Since
+         * we just wrote s0, also make sure to write that same value to
+         * dscratch in case an exception occurs in a program buffer that
+         * might be executed later.
+         */
+        write32(debug_abstract, i++, csrw(S0, CSR_DSCRATCH0));
+      }
 
-        } else {
-          switch (size) {
-            case 2:
-              write32(debug_abstract, i++, fsw(fprnum, ZERO, debug_data_start));
-              break;
-            case 3:
-              write32(debug_abstract, i++, fsd(fprnum, ZERO, debug_data_start));
-              break;
-            default:
-              abstractcs.cmderr = CMDERR_NOTSUP;
-              return true;
-          }
+    } else if (regno >= 0x1020 && regno < 0x1040 && config.support_abstract_fpr_access) {
+      unsigned fprnum = regno - 0x1020;
+
+      if (write) {
+        switch (size) {
+          case 2:
+            write32(debug_abstract, i++, flw(fprnum, ZERO, debug_data_start));
+            break;
+          case 3:
+            write32(debug_abstract, i++, fld(fprnum, ZERO, debug_data_start));
+            break;
+          default:
+            abstractcs.cmderr = CMDERR_NOTSUP;
+            return true;
         }
 
+      } else {
+        switch (size) {
+          case 2:
+            write32(debug_abstract, i++, fsw(fprnum, ZERO, debug_data_start));
+            break;
+          case 3:
+            write32(debug_abstract, i++, fsd(fprnum, ZERO, debug_data_start));
+            break;
+          default:
+            abstractcs.cmderr = CMDERR_NOTSUP;
+            return true;
+        }
+      }
+
       } else if (regno >= 0xc000 && (regno & 1) == 1) {
```

Comment:
```suggestion
      } else if (regno <= 0xffff) {
```

Any reason why only odd-numbered custom CSRs are allowed?

Reply:
wait a minute! this was not in the original scope of changes. I believe that this is require for OpenOCD testing, they had tests targeting this.

Reply:
There is literally the following comment a few lines below:

```
 // Support odd-numbered custom registers, to allow for debugger testing
```

It is worth investigating why this was not detected by the current CI

Reply:
@aap-sc, no, I haven't touched that part. @nadime15, looks like we have justification for why it's done

Reply:
@aap-sc I think thats strange, unless its explicitly specified by the spec. If I define an even-numbered custom register now, I wont be able to access it.

i would like to understand the rationality behind that.

Reply:
@nadime15, please take a look https://github.com/riscv-software-src/riscv-tests/commit/4dddbc79ada7f0a836cf538676c57c8df103ccf6

This is just an arbitrary decision made some time ago by Tim to allow testing both successful and failing accesses.

Reply:
@nadime15 sorry, somehow I've missed your comment. @en-sc already provided the answer.